gcapi/create_strategies.py

```diff
@@ -0,0 +1,328 @@
+        path = Path(s) if isinstance(s, (str, Path)) else None
+        if path and path.exists():
+            cleaned.append(path)
+        else:
+            raise FileNotFoundError(s)
+    if not cleaned:
+        raise FileNotFoundError("No files provided")
+
+    # Ensure no more than can be handled
+    if maximum_number is not None and len(sources) > maximum_number:
+        raise TooManyFiles(
+            f"The maximum is {maximum_number}, "
+            f"you provided {maximum_number}: {sources}"
+        )
+
+    return cleaned
+
+
+class BaseCreateStrategy:
+    """
+    Base class that describes strategies when creating items on Grand Challenge.
+
+    Some items can directly be created via an api call. However, some need
+    a chain of dependent items to exists. The raison d'être of the strategies
+    is to re-use code that checks and handles these relations.
+
+    Each strategy has a `prepare()` function to support batch preparing
+    before things actually land on Grand Challenge. Preparing should **not**
+    create anything on Grand Challenge.
+
+    Calling the strategy will, if applicable, upload contents and hence cause
+    objects to be created on Grand Challenge. Calling the strategy will return
+    a dict that describes how others items can hook up to the created items.
+    """
+
+    def __init__(self, *, client_api):
+        self.client_api = client_api
+        self.prepared = False
+
+    def prepare(self):
+        """Ensure that the strategy can be executed"""
+        self.prepared = True
+        return
+        yield
+
+    def __call__(self):
+        """
+        If applicable, upload any contents and returns an item
+        that can be used in POSTs that would use the created objects.
+        """
+        if not self.prepared:
+            yield from self.prepare()
+
+
+class CIVCreateStrategy(BaseCreateStrategy):
+
+    def __new__(cls, *, interface: ComponentInterface, **__):
+        # Determine the class specialization based on the interface's super_kind.
+        handler_class = {
+            "image": ImageCIVCreateStrategy,
+            "file": FileCIVCreateStrategy,
+            "value": ValueCIVCreateStrategy,
+        }.get(interface.super_kind.casefold())
+
+        if handler_class is None:
+            raise NotImplementedError(
+                f"Unsupported interface super_kind: {interface.super_kind}"
+            )
+
+        if cls is not CIVCreateStrategy:
+            # Specialized class: double check if it supports the interface.
+            if cls is not handler_class:
+                raise RuntimeError(
+                    f"{cls} does not support interface "
+                    f"super_kind: {interface.super_kind}"
+                )
+            return super().__new__(cls)
+
+        return super().__new__(handler_class)
+
+    def __init__(
+        self,
+        *,
+        source: Union[
+            FileSource, HyperlinkedComponentInterfaceValue, HyperlinkedImage
```

Comment:
Do users have a way to reference a `HyperlinkedComponentInterfaceValue`?

Reply:
Yes, via regular gcapi calls that return these. IIRC that is the `DisplaySet`, `ArchiveItem` and `Job` endpoints. Idea is that you can then readily re-use existing CIVs.

Reply:
I think those endpoints just return information about the CIVs (a dictionary with image, file etc., along with the CIV pk). I don't think there is a way to get the `api_url` to a `ComponentInterfaceValue` at the moment, we would need to add that on the GC side.

Reply:
The fun of models.py! =)

https://github.com/DIAGNijmegen/rse-gcapi/blob/0f9b870d0da775221845d3f59ac0449a5678f6d6/gcapi/models.py#L1518

Via DisplaySet model:
https://github.com/DIAGNijmegen/rse-gcapi/blob/0f9b870d0da775221845d3f59ac0449a5678f6d6/gcapi/models.py#L1717

Via:
https://github.com/DIAGNijmegen/rse-gcapi/blob/0f9b870d0da775221845d3f59ac0449a5678f6d6/gcapi/client.py#L149

And finally via:
https://github.com/DIAGNijmegen/rse-gcapi/blob/0f9b870d0da775221845d3f59ac0449a5678f6d6/gcapi/apibase.py#L162

Will end-up being wrapped in that class.

Reply:
Yes, I understand that, but my question was more about how the user will have to supply the CIV to the helper functions. Usually for existing objects (i.e. images) we ask the user to provide the api_url to the image. That does not exist for CIVs yet, and would need to be added on the GC side. But mabye you were envisioning this to work differently?

Reply:
Actually, on second thought, I don't think we'd want to open up an endpoint for CIVs given they contain sensitive information, and we don't have permissions attached to them directly.

Reply:
Main goal of the helper functions is to pull the specifics of CIVs and their instantiations away from the user.

The rational here, for having the typing here it so to allow the user do (pseudo):

```
ds = client.display_set.detail(pk=899...)
client.display_set.create(values=["image-ci-slug": ds.values[0], "float-ci-slug": 42])
```

Ofc they could also do `ds.values[0].image` which works but this opens the hood sort of already.

Perhaps this typing here isn't fully correct and rather it should be typed at `ImageCIVCreateStrategy`, as that specialisation only accepts that. However, the best strategy here is likely supporting this for both file and values as well: File types should download the input, value types should directly use the value.

Thoughts?

Reply:
Ok, it's clear to me now and it makes sense to abstract this away from the user. Ideally this would work for all socket kinds, indeed.

Reply:
I've added handlers for when providing a `HyperlinkedComponentInterfaceValue` as a source.

So one can now source both "file"/"value" super_kind `SocketValue` directly from existing `DisplaySet`/`ArchiveItem`.

Reply:
These strategies are core for the (new) helper functions. With the helper functions steering clear from users having to understand the way the platform handles CIVs, CIs and different super kinds.

As such, I have no clue where I'd warn the users apart from the documentation. Maybe the docstring on the eventual helper functions?

Reply:
Yes, documentation and the docstring of the helper function would do the job, I think...